# General remarks from preprocessing stage

It has been studied previously, that the membership data, the user log data, and the training member data (including the churn flag) contain no duplicates. Duplicates can be found in the transaction data, those have been removed before aggregating information. The user log data is too large to be dealt with in one go using pandas, due to insufficient memory of my laptop, thus aggregation has to be dealt with in chunks in pandas. All of that has been performed in the DataProcessor_KKBox notebook.

In [20]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math

from sklearn import metrics
import datetime as dt
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score, cross_val_predict, RandomizedSearchCV, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, f1_score

import seaborn as sns

import plotly.offline as pyoff
import plotly.graph_objs as go
import plotly.io as pio

## User Activity Time Evolution

Check activity of users by user logs, since we care about churn as inactivity of 30 days, group the user logs in intervals of 30 days. Churn is in fact defined as no renewal of subscription 30 days after a previous membership has been expired. In fact a membership might be ongoing, but a user shows now activity in terms of songs listened to. But user acitivity in terms of played songs can give us an idea about churn, retained or reactivated active users.

In [2]:
df_ul=pd.read_csv("data/user_logs.csv",usecols=['msno','date'])
df_ul["date"]=pd.to_datetime(df_ul["date"], format='%Y%m%d', errors='ignore')
df_ul.head()

msno       date
0  rxIP2f2aN0rYNp+toI0Obt/N/FYQX8hcO1fTmmy2h34= 2015-05-13
1  rxIP2f2aN0rYNp+toI0Obt/N/FYQX8hcO1fTmmy2h34= 2015-07-09
2  yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-01-05
3  yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-03-06
4  yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-05-01

In [3]:
df_ul_v2=pd.read_csv("data/user_logs_V2.csv",usecols=['msno','date'])
df_ul_v2["date"]=pd.to_datetime(df_ul_v2["date"], format='%Y%m%d', errors='ignore')
df_ul_v2

msno       date
0         u9E91QDTvHLq6NXjEaWv8u4QIqhrHk72kE+w31Gnhdg= 2017-03-31
1         nTeWW/eOZA/UHKdD5L7DEqKKFTjaAj3ALLPoAWsU8n0= 2017-03-30
2         2UqkWXwZbIjs03dHLU9KHJNNEvEkZVzm69f3jCS+uLI= 2017-03-31
3         ycwLc+m2O0a85jSLALtr941AaZt9ai8Qwlg9n0Nql5U= 2017-03-31
4         EGcbTofOSOkMmQyN1NMLxHEXJ1yV3t/JdhGwQ9wXjnI= 2017-03-31
...                                                ...        ...
18396357  FGpiy2mB+vXLKziYRcY/xJcJEFJfRDfUqlU+p760f7E= 2017-03-14
18396358  iZRjKNMrw5ffEbfXODLhV/0tJLPbOH3am1WYDgqBf8Q= 2017-03-06
18396359  yztw4Y0EggG0w2wPkbMZx7ke7saSx7dLSfMheHZG/DQ= 2017-03-31
18396360  swCHwkNx30/aENjq30qqaLlm7bUUytbMXdz1bH7g0Jk= 2017-03-07
18396361  pDQSIFZchNPmOproRzaNzy51/8yRXzLK3vvDXKzzGLQ= 2017-03-29

[18396362 rows x 2 columns]

In [4]:
df_ul=df_ul.append(df_ul_v2)
df_ul

msno       date
0         rxIP2f2aN0rYNp+toI0Obt/N/FYQX8hcO1fTmmy2h34= 2015-05-13
1         rxIP2f2aN0rYNp+toI0Obt/N/FYQX8hcO1fTmmy2h34= 2015-07-09
2         yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-01-05
3         yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-03-06
4         yxiEWwE9VR5utpUecLxVdQ5B7NysUPfrNtGINaM2zA8= 2015-05-01
...                                                ...        ...
18396357  FGpiy2mB+vXLKziYRcY/xJcJEFJfRDfUqlU+p760f7E= 2017-03-14
18396358  iZRjKNMrw5ffEbfXODLhV/0tJLPbOH3am1WYDgqBf8Q= 2017-03-06
18396359  yztw4Y0EggG0w2wPkbMZx7ke7saSx7dLSfMheHZG/DQ= 2017-03-31
18396360  swCHwkNx30/aENjq30qqaLlm7bUUytbMXdz1bH7g0Jk= 2017-03-07
18396361  pDQSIFZchNPmOproRzaNzy51/8yRXzLK3vvDXKzzGLQ= 2017-03-29

[410502905 rows x 2 columns]

In [5]:
min_date=df_ul['date'].min()

In [6]:
#df_ul.groupby(["msno",pd.Grouper(key="date",freq="30D", origin='2015-01-01')])

In [7]:
max_date=df_ul['date'].max()

Using the full logging information in terms of active days seems not to work as desired due to high memory consumption while grouping the dates in 30 day intervals, checking on how many days the users had been active. Try now to operate in junks of 180 days, and grouping those per user to reduce the overall load, aka always 6 group periods and then sum up all at the end.

In [8]:
day_period=90
n_time_periods=math.ceil((max_date-min_date)/dt.timedelta(days=day_period))
time_periods=[min_date + i*dt.timedelta(days=day_period) for i in range (0,n_time_periods+1)]
time_periods

[Timestamp('2015-01-01 00:00:00'),
 Timestamp('2015-04-01 00:00:00'),
 Timestamp('2015-06-30 00:00:00'),
 Timestamp('2015-09-28 00:00:00'),
 Timestamp('2015-12-27 00:00:00'),
 Timestamp('2016-03-26 00:00:00'),
 Timestamp('2016-06-24 00:00:00'),
 Timestamp('2016-09-22 00:00:00'),
 Timestamp('2016-12-21 00:00:00'),
 Timestamp('2017-03-21 00:00:00'),
 Timestamp('2017-06-19 00:00:00')]

In [9]:
df_activity_per_user_list=[]
for tp in range(0,len(time_periods)-1):
    print("address periods",time_periods[tp],time_periods[tp+1])
    df_ul_in_period=df_ul[(df_ul['date']>=time_periods[tp]) & (df_ul['date']<time_periods[tp+1])].groupby(["msno",pd.Grouper(
        key="date",freq="30D", origin='2015-01-01')]).agg(
        activedays_per_msno=('msno','count')).reset_index()
    df_activity_per_user_list.append(df_ul_in_period)

address periods 2015-01-01 00:00:00 2015-04-01 00:00:00
address periods 2015-04-01 00:00:00 2015-06-30 00:00:00
address periods 2015-06-30 00:00:00 2015-09-28 00:00:00
address periods 2015-09-28 00:00:00 2015-12-27 00:00:00
address periods 2015-12-27 00:00:00 2016-03-26 00:00:00
address periods 2016-03-26 00:00:00 2016-06-24 00:00:00
address periods 2016-06-24 00:00:00 2016-09-22 00:00:00
address periods 2016-09-22 00:00:00 2016-12-21 00:00:00
address periods 2016-12-21 00:00:00 2017-03-21 00:00:00
address periods 2017-03-21 00:00:00 2017-06-19 00:00:00


In [11]:
large_df_activity_30day_intervalslarge_df_activity_30day_intervalslarge_df_activity_30day_intervals = pd.concat(df_activity_per_user_list, ignore_index=True)
large_df_activity_30day_intervals

msno       date  \
0         +++IZseRRiQS9aaSkH6cMYU6bGDcxUieAi/tH67sC5s= 2015-01-01   
1         +++IZseRRiQS9aaSkH6cMYU6bGDcxUieAi/tH67sC5s= 2015-01-31   
2         +++IZseRRiQS9aaSkH6cMYU6bGDcxUieAi/tH67sC5s= 2015-03-02   
3         +++l/EXNMLTijfLBa8p2TUVVVp2aFGSuUI/h7mLmthw= 2015-01-01   
4         +++l/EXNMLTijfLBa8p2TUVVVp2aFGSuUI/h7mLmthw= 2015-01-31   
...                                                ...        ...   
28702517  zzy0oyiTnRTo5Mbg23oKbBkf9eoaS7+eU4V+d14bzfY= 2017-03-21   
28702518  zzy7iqSpfcRq7R4hmKKuhI+CJRs79a6pteqEggpiNO0= 2017-03-21   
28702519  zzyHq6TK2+cBkeGFUHvh12Z7UxFZiSM7dOOSllSBPDw= 2017-03-21   
28702520  zzz1Dc3P9s53HAowRTrm3fNsWju5yeN4YBfNDq7Z99Q= 2017-03-21   
28702521  zzzF1KsGfHH3qI6qiSNSXC35UXmVKMVFdxkp7xmDMc0= 2017-03-21   

          activedays_per_msno  
0                          20  
1                          15  
2                          30  
3                          12  
4                           9  
...                       ...  
28702517                    8  
28702518                   11  
28702519                   10  
28702520                    5  
28702521                    5  

[28702522 rows x 3 columns]

In [12]:
df_user_act_retention_30days = pd.crosstab(large_df_activity_30day_intervals['msno'], large_df_activity_30day_intervals['date'].dt.date).reset_index()

df_30days_mindate_of_msno = large_df_activity_30day_intervals.groupby('msno')['date'].min().reset_index()
df_30days_mindate_of_msno=df_30days_mindate_of_msno.rename(columns={"date": "min_activity_date"})

df_user_act_retention_30days= pd.merge(df_user_act_retention_30days,df_30days_mindate_of_msno,on="msno")
df_user_act_retention_30days

msno  2015-01-01  2015-01-31  \
0        +++4vcS9aMH7KWdfh5git6nA5fC5jjisd5H/NcM++WM=           0           0   
1        +++EI4HgyhgcJHIPXk/VRP7bt17+2joG39T6oEfJ+tc=           0           0   
2        +++FOrTS7ab3tIgIh8eWwX4FqRv8w/FoiOuyXsFvphY=           0           0   
3        +++IZseRRiQS9aaSkH6cMYU6bGDcxUieAi/tH67sC5s=           1           1   
4        +++TipL0Kt3JvgNE9ahuJ8o+drJAnQINtxD4c5GePXI=           0           0   
...                                               ...         ...         ...   
5339417  zzzqx+aMPSFYjW71JqJ6T/hita+iVemVWzJTE4yQRx8=           0           0   
5339418  zzztPAN9xjMytpZ0RN2gU9mScDULJnHQZK8eZb4uELU=           1           1   
5339419  zzztsqkufVj9DPVJDM3FxDkhlbCL5z4aiYxgPSGkIK4=           0           1   
5339420  zzzueVTwIa5TjXnG2c77bohCVkuksqLkd5mQTP0wTwQ=           0           0   
5339421  zzzyOgMk9MljCerbCCYrVtvu85aSCiy7yCMjAEgNYMs=           0           0   

         2015-03-02  2015-04-01  2015-05-01  2015-05-31  2015-06-30  \
0                 0           1           0           0           0   
1                 0           0           0           0           0   
2                 0           0           0           0           0   
3                 1           1           1           1           1   
4                 0           0           0           0           0   
...             ...         ...         ...         ...         ...   
5339417           0           0           0           0           0   
5339418           1           1           0           0           0   
5339419           1           1           1           0           0   
5339420           0           0           0           0           0   
5339421           0           0           0           1           0   

         2015-07-30  2015-08-29  ...  2016-07-24  2016-08-23  2016-09-22  \
0                 0           0  ...           0           0           0   
1                 0           0  ...           0           0           0   
2                 0           0  ...           0           1           0   
3                 1           1  ...           1           1           1   
4                 0           0  ...           0           0           0   
...             ...         ...  ...         ...         ...         ...   
5339417           0           0  ...           0           0           0   
5339418           0           0  ...           0           0           0   
5339419           0           0  ...           0           0           0   
5339420           0           0  ...           0           0           0   
5339421           0           0  ...           0           0           0   

         2016-10-22  2016-11-21  2016-12-21  2017-01-20  2017-02-19  \
0                 0           0           0           0           0   
1                 0           0           0           0           0   
2                 0           0           0           0           0   
3                 1           1           1           1           1   
4                 0           0           0           0           0   
...             ...         ...         ...         ...         ...   
5339417           0           0           0           0           0   
5339418           0           0           0           0           0   
5339419           0           0           0           0           0   
5339420           0           0           0           0           1   
5339421           0           0           0           0           0   

         2017-03-21  min_activity_date  
0                 0         2015-04-01  
1                 0         2016-03-26  
2                 0         2016-08-23  
3                 1         2015-01-01  
4                 0         2015-12-27  
...             ...                ...  
5339417           0         2016-02-25  
5339418           0         2015-01-01  
5339419           0         2015-01-31  
5339420           0         2017-02-19  
533942

In [13]:
check_dates_30d = df_user_act_retention_30days.columns[1:-1]
print(check_dates_30d)

Index([2015-01-01, 2015-01-31, 2015-03-02, 2015-04-01, 2015-05-01, 2015-05-31,
       2015-06-30, 2015-07-30, 2015-08-29, 2015-09-28, 2015-10-28, 2015-11-27,
       2015-12-27, 2016-01-26, 2016-02-25, 2016-03-26, 2016-04-25, 2016-05-25,
       2016-06-24, 2016-07-24, 2016-08-23, 2016-09-22, 2016-10-22, 2016-11-21,
       2016-12-21, 2017-01-20, 2017-02-19, 2017-03-21],
      dtype='object')


In [15]:
retention_array = []
for i in range(len(check_dates_30d)-1):
    retention_data = {}
    selected_check_date = check_dates_30d[i+1]
    prev_check_date = check_dates_30d[i]
    retention_data['min_activity_date'] = selected_check_date
    retention_data['TotalActiveUserCount'] = df_user_act_retention_30days[selected_check_date].sum()
    retention_data['NewUserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (selected_check_date==df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].sum()
    retention_data['RetainedUserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]>0)][selected_check_date].sum()
    retention_data['InactiveUserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]==0) & (df_user_act_retention_30days[prev_check_date]>0)][selected_check_date].count()
    #Customers reactivated after at least 1 period of inactivity
    retention_data['Reactivated_1Period_PLUS_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    # Customers reactivated after at least 2 periods of inactivity, we can only check that if we are two periods in
    if (i>0):
        second_prev_check_date = check_dates_30d[i-1] # if months[i-1].isna() == False else 0
        retention_data['Reactivated_2Period_PLUS_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (df_user_act_retention_30days[second_prev_check_date]==0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    else:
        retention_data['Reactivated_2Period_PLUS_UserCount'] = 0
    # Customers reactivated after at least 3 periods of inactivity
    if (i>1):
        second_prev_check_date = check_dates_30d[i-1] # if months[i-1].isna() == False else 0
        third_prev_check_date = check_dates_30d[i-2] # if months[i-2].isna() == False else 0
        retention_data['Reactivated_3Period_PLUS_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (df_user_act_retention_30days[second_prev_check_date]==0) & (df_user_act_retention_30days[third_prev_check_date]==0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    else:
        retention_data['Reactivated_3Period_PLUS_UserCount'] = 0
        
    # Customers reactivated after at ONLY 1 period of inactivity
    if(i==0):
        retention_data['Reactivated_1Period_ONLY_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) &  (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    else: 
        retention_data['Reactivated_1Period_ONLY_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (df_user_act_retention_30days[second_prev_check_date]>0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count() 
    
    # Customers reactivated after at ONLY 2 periods of inactivity
    if(i>0):
        second_prev_check_date = check_dates_30d[i-1] # if months[i-1].isna() == False else 0
        retention_data['Reactivated_2Period_ONLY_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (df_user_act_retention_30days[second_prev_check_date]==0) & (df_user_act_retention_30days[third_prev_check_date]>0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    else:
        retention_data['Reactivated_2Period_ONLY_UserCount'] = 0
    # Customers reactivated after at ONLY 3 periods of inactivity
    third_prev_check_date = check_dates_30d[i-2] # if months[i-2].isna() == False else 0
    if(i>2):
        second_prev_check_date = check_dates_30d[i-1] # if months[i-1].isna() == False else 0
        fourth_prev_check_date = check_dates_30d[i-3] # if months[i-2].isna() == False else 0
        third_prev_check_date = check_dates_30d[i-2] # if months[i-2].isna() == False else 0
        retention_data['Reactivated_3Period_ONLY_UserCount'] = df_user_act_retention_30days[(df_user_act_retention_30days[selected_check_date]>0) & (df_user_act_retention_30days[prev_check_date]==0) & (df_user_act_retention_30days[second_prev_check_date]==0) & (df_user_act_retention_30days[third_prev_check_date]==0) & (prev_check_date>df_user_act_retention_30days['min_activity_date'].dt.date)][selected_check_date].count()
    else:
        retention_data['Reactivated_3Period_ONLY_UserCount'] = 0
    retention_array.append(retention_data)

In [16]:
tx_retention_plot = pd.DataFrame(retention_array)
tx_retention_plot["TotalActiveUserCount_prevPeriod"]=tx_retention_plot["TotalActiveUserCount"].shift(1)
tx_retention_plot["ActiveUserRate"]=tx_retention_plot["TotalActiveUserCount"]/tx_retention_plot["TotalActiveUserCount_prevPeriod"]
tx_retention_plot["NewUserRate"]=tx_retention_plot["NewUserCount"]/tx_retention_plot["TotalActiveUserCount_prevPeriod"]
tx_retention_plot["RetainedUserRate"]=tx_retention_plot["RetainedUserCount"]/tx_retention_plot["TotalActiveUserCount_prevPeriod"]
tx_retention_plot["ChurnRate"]=tx_retention_plot["InactiveUserCount"]/tx_retention_plot["TotalActiveUserCount_prevPeriod"]
tx_retention_plot["ReactivationRate"]=tx_retention_plot["Reactivated_1Period_PLUS_UserCount"]/tx_retention_plot["TotalActiveUserCount_prevPeriod"]
#print(np.max(tx_retention_plot["TotalActiveUserCount"]-tx_retention_plot["NewUserCount"]-tx_retention_plot["RetainedUserCount"]-tx_retention_plot["Reactivated_1Period_PLUS_UserCount"]))
tx_retention_plot.head(50)

min_activity_date  TotalActiveUserCount  NewUserCount  RetainedUserCount  \
0         2015-01-31                948986        210603             738383   
1         2015-03-02                936596        172092             739768   
2         2015-04-01                939930        145035             751170   
3         2015-05-01                917844        118848             754980   
4         2015-05-31                918345        116150             751769   
5         2015-06-30                866356        119409             713611   
6         2015-07-30                907676        167353             703728   
7         2015-08-29                907009        137981             730609   
8         2015-09-28                992513        214981             742114   
9         2015-10-28               1039647        242735             761638   
10        2015-11-27               1031767        216285             778852   
11        2015-12-27               1063238        230213             795314   
12        2016-01-26               1048244        209721             802637   
13        2016-02-25               1046477        185783             822741   
14        2016-03-26               1044692        168665             839002   
15        2016-04-25               1044244        161749             846807   
16        2016-05-25               1067073        173835             855201   
17        2016-06-24               1097448        183561             869337   
18        2016-07-24               1091889        169572             885113   
19        2016-08-23               1114358        161304             895876   
20        2016-09-22               1100099        152688             904822   
21        2016-10-22               1183203        178202             913910   
22        2016-11-21               1139352        143927             941372   
23        2016-12-21               1118958        134682             936094   
24        2017-01-20               1131298        143272             943138   
25        2017-02-19               1120778        125589             954398   
26        2017-03-21                954765         25450             910990   

    InactiveUserCount  Reactivated_1Period_PLUS_UserCount  \
0              191354                                   0   
1              209218                               24736   
2              185426                               43725   
3              184950                               44016   
4              166075                               50426   
5              204734                               33336   
6              162628                               36595   
7              177067                               38419   
8              164895                               35418   
9              230875                               35274   
10             260795                               36630   
11             236453                               37711   
12             260601                               35886   
13             225503                               37953   
14             207475                               37025   
15             197885                               35688   
16             189043                               38037   
17             197736                               44550   
18             212335                               37204   
19             196013                               57178   
20             209536                               42589   
21             186189                               91091   
22             241831                               54053   
23             203258                               48182   
24             175820                               44888   
25             176900                               40791   
26             209788                               18325   

    Reactivated_2Period_PLUS_UserCount  Reactivated_3Period_PLUS_UserCount  \
0  

In [24]:
from datetime import datetime, timedelta,date

tx_retention_plot=tx_retention_plot[(tx_retention_plot["min_activity_date"]>datetime.strptime("2015-01-01", '%Y-%m-%d').date())
                                & (tx_retention_plot["min_activity_date"]<datetime.strptime("2017-03-21", '%Y-%m-%d').date())]

In [26]:
plot_data = [
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['TotalActiveUserCount'],
        name="TotalActiveUserCount"
    ),
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['NewUserCount'],
        name="NewUserCount"
    ),
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['RetainedUserCount'],
        name="RetainedUserCount"
    ),
    # Inactive customers keep adding up as we know from Recency, therefore this is not shown to avoid squeezing the chart
    go.Scatter(
       x=tx_retention_plot['min_activity_date'],
       y=tx_retention_plot['InactiveUserCount'],
       name="ChurnedUserCount"
    ),
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['Reactivated_1Period_PLUS_UserCount'],
        name="ReactivatedUserCount"
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Periodical (30 days) Users Evolution'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)
pio.write_image(fig, 'figures/KKBox_30d_UserEvolution_AllAbsoluteNumbers.png')

In [30]:
tx_retention_plot=tx_retention_plot[(tx_retention_plot["min_activity_date"]>datetime.strptime("2015-01-31", '%Y-%m-%d').date())
                                & (tx_retention_plot["min_activity_date"]<datetime.strptime("2017-03-21", '%Y-%m-%d').date())]

In [33]:
plot_data = [

    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['RetainedUserRate'],
        name="Retained User Rate"
    ),
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['ReactivationRate'],
        name="Reactivated User Rate"
    ),
    go.Scatter(
        x=tx_retention_plot['min_activity_date'],
        y=tx_retention_plot['ChurnRate'],
        name="Churn Rate"
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Periodical (30 days) Users Evolution',
        yaxis={"tickformat": ',.0%',
                "range": [0,1.0]}
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.update_layout(
    font=dict(size=16),
    legend=dict(
    #title_text='Rate To Previous Active Users',
    yanchor="top",
    y=0.75,
    xanchor="right",
    x=0.80
))
pyoff.iplot(fig)
pio.write_image(fig, 'figures/KKBox_30d_UserEvolution_30d_UserEvolution_Retained_Reactivation_Rates.png')

### Overall User Evolution

Studied in periods of 30 days the retained user rate is rather flat around 80 %, the reactivated user rate is typically of the order of 3-4 %, thus the churn rate is around 20 %, slightly lower in the first quarter of 2017. Definitions are defined as activity in terms of users streaming songs.

In [34]:
df_mem=pd.read_csv("data/members_v3.csv")
df_mem

msno  city  bd  gender  \
0        Rb9UwLQTrxzBVwCB6+bCcSQWZ9JiNLC9dXtM1oEsZA8=     1   0     NaN   
1        +tJonkh+O1CA796Fm5X60UMOtB6POHAwPjbTRVl/EuU=     1   0     NaN   
2        cV358ssn7a0f7jZOwGNWS07wCKVqxyiImJUX6xcIwKw=     1   0     NaN   
3        9bzDeJP6sQodK73K5CBlJ6fgIQzPeLnRl0p5B77XP+g=     1   0     NaN   
4        WFLY3s7z4EZsieHCt63XrsdtfTEmJ+2PnnKLH5GY4Tk=     6  32  female   
...                                               ...   ...  ..     ...   
6769468  VSGkb3hyBRUtb/b1MQUZbvOkktS3vKLnhMHW0CF8eyU=     1   0     NaN   
6769469  nWjH7glPkZ7jOVaCRwwjlpmp0T1hSWdv8hMJxiWCwKc=     1   0     NaN   
6769470  GH+b5+1tlv7ZZXsA8upBzVXMTLyffKcsF7WoU8b5rOI=    15  26  female   
6769471  XVlwT3fdCFGKqerEKBzUIjK+jzI6jzSke4cDMVhYyjE=     1   0     NaN   
6769472  isBt+JlgvZRNy6lxVr5vvuJ4lD00ofTaKyJ+uRnGcPg=     1   0     NaN   

         registered_via  registration_init_time  
0                    11                20110911  
1                     7                20110914  
2                    11                20110915  
3                    11                20110915  
4                     9                20110915  
...                 ...                     ...  
6769468               7                20151020  
6769469               7                20151020  
6769470               4                20151020  
6769471               4                20151020  
6769472               4                20151021  

[6769473 rows x 6 columns]

In [51]:
df_mem["registration_init_time_as_dt"]=pd.to_datetime(df_mem["registration_init_time"], format='%Y%m%d', errors='ignore')
df_mem.sort_values(by="registration_init_time")

msno  city  bd  gender  \
2778759  +1VIk9gpQ9cOLXPbQOxeLm2VjH9oKWtHyIMwe+BfFt0=    21  32    male   
6363176  yCQADf9WWjAUSBRJHtZH31WrmL6MsxpBo165UMpUA6I=    10   0  female   
82956    XT0yF1DRjbn5ymbnsasJnag4+53huda0TZ3bRPIcrAA=    13  32    male   
107040   OVR9+ssQRSxtzfQjcVtdoiZ6c4bHIuLdhUTVdA9oR0M=    13  30  female   
107041   3kM1IV7sTY3OIPNt/8IG1Ahwlg57T7/BOL1VzVUB4HM=     4  39    male   
...                                               ...   ...  ..     ...   
5152046  Np0Xkl+nRyBKgIZh1z77he6Fis27DNQTk8F5IJc4Rn8=     1  40  female   
5152045  xl587vDc3ZhkRvA+hk9PI7B8MOZxDBrikbOsou5w3Nc=     1   0     NaN   
5152044  FN00EIM5kDR3aR8rRoRVWlBIHoNZSdlHSgOSSzfL5ng=     1  46    male   
252864   8hYHwm6tMWzkqBd+26xr1QfcuHrSpq63Cj2IaB7XGFc=     1  21    male   
6387922  7Ful/iRpY6+/2BjnlNQjsDBbQ8h82zBL7lG4m2fMMtQ=     1   0     NaN   

         registered_via  registration_init_time registration_init_time_as_dt  
2778759               9                20040326                   2004-03-26  
6363176               9                20040326                   2004-03-26  
82956                 9                20040326                   2004-03-26  
107040                9                20040326                   2004-03-26  
107041                9                20040326                   2004-03-26  
...                 ...                     ...                          ...  
5152046               3                20170429                   2017-04-29  
5152045               7                20170429                   2017-04-29  
5152044               3                20170429                   2017-04-29  
252864                3                20170429                   2017-04-29  
6387922               9                20170429                   2017-04-29  

[6769473 rows x 7 columns]

In [35]:
print("age 0",len(df_mem[df_mem["bd"]==0]),"neg age",len(df_mem[df_mem["bd"]<0]), "age 0<10",len(df_mem[(df_mem["bd"]>0) & (df_mem["bd"]<10)]),
     "age 10-20", len(df_mem[(df_mem["bd"]>=10) & (df_mem["bd"]<20)]),"age 20-50",len(df_mem[(df_mem["bd"]>=20) & (df_mem["bd"]<50)]),
     "age 50-100",len(df_mem[(df_mem["bd"]>=50) & (df_mem["bd"]<100)]),"age >100",len(df_mem[df_mem["bd"]>100]))

age 0 4540215 neg age 274 age 0<10 899 age 10-20 330391 age 20-50 1779048 age 50-100 113126 age >100 5377


In [55]:
1937764/2426143

0.7987014780249969

Age (bd column) offers no useful information as well, for about two thirds of customers the age is 0, also impossible ages are listed, thus will ignore that information for now. For 5 339 422 out of the 6 769 473 users we have the user log information available. Only for 2 425 143 users we have transaction data available.

After merging the user log data and the membership data we have 5 169 234 members left (96.8 % of the member for which we have user logs), after merging of the transaction data we are left with 79.9 % of the original transaction data. In user log overall data we have a couple of users which have 1 active day, for those standard deviations of distributions is not defined properly.

It has been checked that both transaction data and user log data contain information starting from Jan 1, 2015 to March 31, 2017.

In [53]:
df_user_log_agg_read=pd.read_csv("df_user_log_aggregated_per_user_def.csv")
df_user_log_agg_read=df_user_log_agg_read.drop(columns="Unnamed: 0")
df_user_log_agg_read["last_date_active"]=pd.to_datetime(df_user_log_agg_read["last_date_active"], format='%Y-%m-%d', errors='ignore')
print(df_user_log_agg_read.dtypes)
df_user_log_agg_read

msno                           object
activity_duration               int64
last_date_active       datetime64[ns]
n_active_days                   int64
sum_num_25                      int64
sum_num_50                      int64
sum_num_75                      int64
sum_num_985                     int64
sum_num_100                     int64
sum_num_unq                     int64
tot_secs                      float64
mean_num_25                   float64
median_num_25                 float64
q25_num_25                    float64
q75_num_25                    float64
std_num_25                    float64
median_num_50                 float64
std_num_50                    float64
median_num_75                 float64
std_num_75                    float64
median_num_985                float64
std_num_985                   float64
mean_num_100                  float64
median_num_100                float64
q25_num_100                   float64
q75_num_100                   float64
std_num_100 

msno  activity_duration  \
0        +++FOrTS7ab3tIgIh8eWwX4FqRv8w/FoiOuyXsFvphY=                  6   
1        +++dz9ZCWE2HB/47pJU82NJXQzQuZDx1Wm50YSk/kKk=                  3   
2        ++5IYGT7+CWMJ8hRsqoQLoaTXBLMJzlfK12eMTr9Ilw=                118   
3        ++7jYuHyUSp41PyuttFx/MCepv7TdFQULgN8TxZULZk=                542   
4        ++92FghbCQPqDmQ96QzNiuEMoDxrMOmuaisu1UCrYn0=                347   
...                                               ...                ...   
5339417  zzjrmRFJWpHm5f6OcUTHhnj3Xp6rQhaVC4isJBbjm+8=                 14   
5339418  zzpujC8UYr82W9wURtuMOI+hCaskgYEluKz1TJKFoEY=                  2   
5339419  zzu6u0H6kxDzwhPY7KpdvV4CbMsjIbBVSM6fYXwTrE4=                  0   
5339420  zzxIjlrMxRzNVDD6XqUDRBqkWn7OJ9mzFmFQlfLtLCE=                 26   
5339421  zzzueVTwIa5TjXnG2c77bohCVkuksqLkd5mQTP0wTwQ=                  0   

        last_date_active  n_active_days  sum_num_25  sum_num_50  sum_num_75  \
0             2016-09-15              7          60          12          14   
1             2016-03-02              4           5           1           1   
2             2015-05-01             63           9           5           0   
3             2016-06-30            183         669         115          82   
4             2017-01-22            279         465         138         126   
...                  ...            ...         ...         ...         ...   
5339417       2017-03-24              8          22           6           2   
5339418       2017-03-16              3           4           0           0   
5339419       2017-03-09              1           7           2           0   
5339420       2017-03-31             12          11          10           3   
5339421       2017-03-03              1           2           0           0   

         sum_num_985  sum_num_100  sum_num_unq  ...  n_max_break  n_min_break  \
0                  7          171          179  ...          0.0          0.0   
1                  1          183          147  ...          0.0          0.0   
2                  3          349          191  ...          5.0          0.0   
3                 79         1790         2250  ...         25.0          0.0   
4                143        10690        10699  ...          6.0          0.0   
...              ...          ...          ...  ...          ...          ...   
5339417            1          135          157  ...          4.0          0.0   
5339418            0          149          112  ...          0.0          0.0   
5339419            1           11           11  ...          NaN          NaN   
5339420            3           27           38  ...          7.0          0.0   
5339421            0            1            3  ...          NaN          NaN   

         n_mean_break  n_mode_break  n_median_break  n_max_break_last  \
0            0.000000           0.0             0.0               0.0   
1            0.000000           0.0             0.0               0.0   
2            0.903226           0.0             0.0               5.0   
3            1.978022           0.0             1.0              16.0   
4            0.248201           0.0             0.0               6.0   
...               ...           ...             ...               ...   
5339417      1.000000           0.0             1.0               4.0   
5339418      0.000000           0.0             0.0               0.0   
5339419           NaN           NaN             NaN               NaN   
5339420      1.363636           0.0             0.0               7.0   
5339421           NaN           NaN             NaN               NaN   

         n_min_break_last  n_mean_break_last  n_mode_break_last  \
0                     0.0           0.000000                0.0   
1                     0.0           0.000000                0.0   
2                     0.0           2.444444                0.0   
3                    14.0          15.000000               14.0   
4 

In [43]:
"""
df_tra=pd.read_csv("data/transactions.csv")
df_tra_V2=pd.read_csv("data/transactions_v2.csv")
df_tra=df_tra.append(df_tra_V2)
df_tra=df_tra.drop_duplicates(ignore_index=True)
df_tra['msno'].nunique()
"""

2426143

In [48]:
df_tra_agg_per_user=pd.read_csv("df_transaction_aggregated_per_user_def_with_duration_NEW.csv")
df_tra_agg_per_user=df_tra_agg_per_user.drop(columns="Unnamed: 0")
df_tra_agg_per_user["max_date_trans"]=pd.to_datetime(df_tra_agg_per_user["max_date_trans"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["min_date_trans"]=pd.to_datetime(df_tra_agg_per_user["min_date_trans"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["max_memb_exp"]=pd.to_datetime(df_tra_agg_per_user["max_memb_exp"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["first_cancel_membership_expire_date"]=pd.to_datetime(df_tra_agg_per_user["first_cancel_membership_expire_date"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["last_cancel_membership_expire_date"]=pd.to_datetime(df_tra_agg_per_user["last_cancel_membership_expire_date"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["first_auto_renew_transaction"]=pd.to_datetime(df_tra_agg_per_user["first_auto_renew_transaction"], format='%Y-%m-%d', errors='ignore')
df_tra_agg_per_user["last_auto_renew_transaction"]=pd.to_datetime(df_tra_agg_per_user["last_auto_renew_transaction"], format='%Y-%m-%d', errors='ignore')
print(df_tra_agg_per_user.dtypes)
df_tra_agg_per_user


msno                                                      object
total_num_transaction                                      int64
max_date_trans                                    datetime64[ns]
min_date_trans                                    datetime64[ns]
max_memb_exp                                      datetime64[ns]
min_amount_paid                                            int64
max_amount_paid                                            int64
mean_amount_paid                                         float64
median_amount_paid                                       float64
mode_amount_paid                                           int64
total_amount_paid                                          int64
total_plan_list_price                                      int64
mean_plan_list_price                                     float64
median_plan_list_price                                   float64
mode_plan_list_price                                       int64
num_payment_methods      

msno  total_num_transaction  \
0        +++FOrTS7ab3tIgIh8eWwX4FqRv8w/FoiOuyXsFvphY=                      1   
1        +++IZseRRiQS9aaSkH6cMYU6bGDcxUieAi/tH67sC5s=                      2   
2        +++hVY1rZox/33YtvDgmKA2Frg/2qhkz12B9ylCvh8o=                      5   
3        +++l/EXNMLTijfLBa8p2TUVVVp2aFGSuUI/h7mLmthw=                     21   
4        +++snpr7pmobhLKUgSHTv/mpkqgBT0tQJ0zQj6qKrqc=                     27   
...                                               ...                    ...   
2426138  zzz9+ZF4+GMyt63oU8xfjo1EkvRqH5OINlES0RUJI6I=                     14   
2426139  zzzF1KsGfHH3qI6qiSNSXC35UXmVKMVFdxkp7xmDMc0=                      2   
2426140  zzzN9thH22os1dRS0VHReY/8FTfGHOi86//d+wGGFsQ=                     12   
2426141  zzztsqkufVj9DPVJDM3FxDkhlbCL5z4aiYxgPSGkIK4=                      1   
2426142  zzzyOgMk9MljCerbCCYrVtvu85aSCiy7yCMjAEgNYMs=                      1   

        max_date_trans min_date_trans max_memb_exp  min_amount_paid  \
0           2016-09-09     2016-09-09   2016-09-14                0   
1           2016-10-23     2015-11-21   2018-02-06             1599   
2           2017-03-15     2016-11-16   2017-04-15               99   
3           2017-03-31     2015-01-31   2017-05-19              149   
4           2017-03-26     2015-01-26   2017-04-26              149   
...                ...            ...          ...              ...   
2426138     2016-11-03     2015-10-14   2016-11-13              129   
2426139     2017-03-05     2017-02-05   2017-04-04              149   
2426140     2017-02-02     2016-03-04   2017-02-04               99   
2426141     2015-06-08     2015-06-08   2015-06-15                0   
2426142     2015-06-13     2015-06-13   2015-06-15                0   

         max_amount_paid  mean_amount_paid  median_amount_paid  \
0                      0          0.000000                 0.0   
1                   1788       1693.500000              1693.5   
2                     99         99.000000                99.0   
3                    149        149.000000               149.0   
4                    149        149.000000               149.0   
...                  ...               ...                 ...   
2426138              149        131.857143               129.0   
2426139              149        149.000000               149.0   
2426140               99         99.000000                99.0   
2426141                0          0.000000                 0.0   
2426142                0          0.000000                 0.0   

         mode_amount_paid  ...  median_delta_trans_date  \
0                       0  ...                      NaN   
1                    1599  ...                    337.0   
2                      99  ...                     30.0   
3                     149  ...                     31.0   
4                     149  ...                     31.0   
...                   ...  ...                      ...   
2426138               129  ...                     30.0   
2426139               149  ...                     28.0   
2426140                99  ...                     31.0   
2426141                 0  ...                      NaN   
2426142                 0  ...                      NaN   

         mode_delta_trans_date  min_delta_amount_paid  max_delta_amount_paid  \
0                          NaN                    NaN                    NaN   
1                        337.0                 -189.0                 -189.0   
2                         31.0                    0.0                    0.0   
3                         31.0                    0.0                    0.0   
4                         31.0                    0.0                    0.0   
...                        ...                    ...                    ...   
2426138                   31.0                    0.0                   20.0   
2426139                   28.0                    0.0                    0.0   
2426140      

In [60]:
df_user_all_info_agg=df_mem.merge(df_user_log_agg_read,on="msno")
print("after membership merge",len(df_user_all_info_agg))
df_user_all_info_agg=df_user_all_info_agg.merge(df_tra_agg_per_user,on="msno")
print("after transaction merge",len(df_user_all_info_agg))
df_user_all_info_agg

after membership merge 5169234
after transaction merge 1937764


msno  city  bd  gender  \
0        +tJonkh+O1CA796Fm5X60UMOtB6POHAwPjbTRVl/EuU=     1   0     NaN   
1        WFLY3s7z4EZsieHCt63XrsdtfTEmJ+2PnnKLH5GY4Tk=     6  32  female   
2        yLkV2gbZ4GLFwqTOXLVHz0VGrMYcgBGgKZ3kj9RiYu8=     4  30    male   
3        WH5Jq4mgtfUFXh2yz+HrcTXKS4Oess4k4W3qKolAeb0=     5  34    male   
4        I0yFvqMoNkM8ZNHb617e1RBzIS/YRKemHO7Wj13EtA0=    13  63    male   
...                                               ...   ...  ..     ...   
1937759  OL6joplCKThE1pZD+AQDnj3kZiC45qzpndlDIRSyPQI=     1   0     NaN   
1937760  O11rAa17ANM6DGMMVb9L2vwCQFv/S6SR3SDB1nGuOAc=     1   0     NaN   
1937761  tyoRqQ72Y2oCcqvdaIxG8T0Hs9DZvuuM8IeTwhxtdBQ=     6  26  female   
1937762  VSGkb3hyBRUtb/b1MQUZbvOkktS3vKLnhMHW0CF8eyU=     1   0     NaN   
1937763  nWjH7glPkZ7jOVaCRwwjlpmp0T1hSWdv8hMJxiWCwKc=     1   0     NaN   

         registered_via  registration_init_time registration_init_time_as_dt  \
0                     7                20110914                   2011-09-14   
1                     9                20110915                   2011-09-15   
2                     9                20110916                   2011-09-16   
3                     9                20110916                   2011-09-16   
4                     9                20110918                   2011-09-18   
...                 ...                     ...                          ...   
1937759               7                20151020                   2015-10-20   
1937760               7                20151020                   2015-10-20   
1937761               7                20151020                   2015-10-20   
1937762               7                20151020                   2015-10-20   
1937763               7                20151020                   2015-10-20   

         activity_duration last_date_active  n_active_days  ...  \
0                      820       2017-03-31            763  ...   
1                       14       2017-03-28              9  ...   
2                      781       2017-03-31             18  ...   
3                      220       2017-03-30             73  ...   
4                      817       2017-03-30            354  ...   
...                    ...              ...            ...  ...   
1937759                528       2017-03-31            396  ...   
1937760                525       2017-03-31            273  ...   
1937761                528       2017-03-31            417  ...   
1937762                481       2017-02-15              2  ...   
1937763                518       2017-03-26            188  ...   

         median_delta_trans_date  mode_delta_trans_date  \
0                           31.0                   31.0   
1                            NaN                    NaN   
2                           31.0                   31.0   
3                          179.0                  179.0   
4                           31.0                   31.0   
...                          ...                    ...   
1937759                     31.0                   31.0   
1937760                     31.0                   31.0   
1937761                     30.5                   31.0   
1937762                     26.0                   26.0   
1937763                     31.0                   31.0   

         min_delta_amount_paid  max_delta_amount_paid  mode_delta_amount_paid  \
0                        -20.0                   20.0                     0.0   
1                          NaN                    NaN                     NaN   
2                          0.0                    0.0                     0.0   
3                          0.0                    0.0                     0.0   
4                          0.0                    0.0                     0.0   
...                        ...                    ...                     ...   
1937759                    0.0                    0.0                     0.0   
1937760                  

In [65]:
for col in df_user_all_info_agg.columns:
    print(col,df_user_all_info_agg[col].isna().sum())

msno 0
city 0
bd 0
gender 958307
registered_via 0
registration_init_time 0
registration_init_time_as_dt 0
activity_duration 0
last_date_active 0
n_active_days 0
sum_num_25 0
sum_num_50 0
sum_num_75 0
sum_num_985 0
sum_num_100 0
sum_num_unq 0
tot_secs 0
mean_num_25 0
median_num_25 0
q25_num_25 0
q75_num_25 0
std_num_25 87507
median_num_50 0
std_num_50 87507
median_num_75 0
std_num_75 87507
median_num_985 0
std_num_985 87507
mean_num_100 0
median_num_100 0
q25_num_100 0
q75_num_100 0
std_num_100 87507
mean_tot_secs 0
median_tot_secs 0
q25_tot_secs 0
q75_tot_secs 0
std_tot_secs 87507
n_max_break 87507
n_min_break 87507
n_mean_break 87507
n_mode_break 87507
n_median_break 87507
n_max_break_last 87507
n_min_break_last 87507
n_mean_break_last 87507
n_mode_break_last 87507
n_median_break_last 87507
total_num_transaction 0
max_date_trans 0
min_date_trans 0
max_memb_exp 0
min_amount_paid 0
max_amount_paid 0
mean_amount_paid 0
median_amount_paid 0
mode_amount_paid 0
total_amount_paid 0
total_pla

In [70]:
df_user_all_info_agg[df_user_all_info_agg["std_tot_secs"].isna()]["mean_tot_secs"].max()

1271346.5259999998

In [63]:
df_user_all_info_agg_outer=df_mem.merge(df_user_log_agg_read,on="msno",how="outer")
print("after membership merge_outer",len(df_user_all_info_agg_outer))
df_user_all_info_ag_outerg=df_user_all_info_agg_outer.merge(df_tra_agg_per_user,on="msno",how="outer")
print("after transaction merge_outer",len(df_user_all_info_ag_outerg))
df_user_all_info_agg_outer

after membership merge_outer 6939661
after transaction merge_outer 7377004


msno  city    bd  gender  \
0        Rb9UwLQTrxzBVwCB6+bCcSQWZ9JiNLC9dXtM1oEsZA8=   1.0   0.0     NaN   
1        +tJonkh+O1CA796Fm5X60UMOtB6POHAwPjbTRVl/EuU=   1.0   0.0     NaN   
2        cV358ssn7a0f7jZOwGNWS07wCKVqxyiImJUX6xcIwKw=   1.0   0.0     NaN   
3        9bzDeJP6sQodK73K5CBlJ6fgIQzPeLnRl0p5B77XP+g=   1.0   0.0     NaN   
4        WFLY3s7z4EZsieHCt63XrsdtfTEmJ+2PnnKLH5GY4Tk=   6.0  32.0  female   
...                                               ...   ...   ...     ...   
6939656  PQBDtyEUkJdv7Zb0ebClh0WhH/WIu11XQNU5sEcetP8=   NaN   NaN     NaN   
6939657  QQapErUBXq59Q+8chH1yTEzEpfT4YfRx9R/P2ToYXaA=   NaN   NaN     NaN   
6939658  QuCXSKyV44OS+L4ofixj9QUgSrwB7DSQRd/YWhHL1hg=   NaN   NaN     NaN   
6939659  gmhNhbDfUgBur7/pwg8xLdlTOFzRVwtsJZsi6tdrAxE=   NaN   NaN     NaN   
6939660  yFuyvlPZwLu3GiWFzylKnwrjQifdePUkcUq7sIc6pps=   NaN   NaN     NaN   

         registered_via  registration_init_time registration_init_time_as_dt  \
0                  11.0              20110911.0                   2011-09-11   
1                   7.0              20110914.0                   2011-09-14   
2                  11.0              20110915.0                   2011-09-15   
3                  11.0              20110915.0                   2011-09-15   
4                   9.0              20110915.0                   2011-09-15   
...                 ...                     ...                          ...   
6939656             NaN                     NaN                          NaT   
6939657             NaN                     NaN                          NaT   
6939658             NaN                     NaN                          NaT   
6939659             NaN                     NaN                          NaT   
6939660             NaN                     NaN                          NaT   

         activity_duration last_date_active  n_active_days  ...  n_max_break  \
0                      NaN              NaT            NaN  ...          NaN   
1                    820.0       2017-03-31          763.0  ...          7.0   
2                      NaN              NaT            NaN  ...          NaN   
3                      NaN              NaT            NaN  ...          NaN   
4                     14.0       2017-03-28            9.0  ...          3.0   
...                    ...              ...            ...  ...          ...   
6939656                0.0       2017-03-03            1.0  ...          NaN   
6939657                0.0       2017-03-19            1.0  ...          NaN   
6939658                1.0       2017-03-31            2.0  ...          0.0   
6939659                0.0       2017-03-05            1.0  ...          NaN   
6939660                1.0       2017-03-31            2.0  ...          0.0   

         n_min_break  n_mean_break  n_mode_break  n_median_break  \
0                NaN           NaN           NaN             NaN   
1                0.0      0.076115           0.0             0.0   
2                NaN           NaN           NaN             NaN   
3                NaN           NaN           NaN             NaN   
4                0.0      0.750000           0.0             0.0   
...              ...           ...           ...             ...   
6939656          NaN           NaN           NaN             NaN   
6939657          NaN           NaN           NaN             NaN   
6939658          0.0      0.000000           0.0             0.0   
6939659          NaN           NaN           NaN             NaN   
6939660          0.0      0.000000           0.0             0.0   

         n_max_break_last  n_min_break_last  n_mean_break_last  \
0                     NaN               NaN                NaN   
1                     1.0               0.0           0.034483   
2                     NaN               NaN                NaN   
3                     NaN               NaN                NaN   
4                     3.0               0.0           0

In [72]:
df_train=pd.read_csv("data/train.csv")
df_train.sort_values(by=["msno"])
df_train.nunique()

msno        992931
is_churn         2
dtype: int64

In [73]:
df_train_v2=pd.read_csv("data/train_v2.csv")
df_train_v2.sort_values(by=["msno"])
df_train_v2.nunique()

msno        970960
is_churn         2
dtype: int64

Out of 970 960 users in train dataset, 855 160 are available in the user log dataset, 860 967 are available in the membership dataset, 855 136 are available in each of the user log, transaction, and membership data. All of training data is available in the transaction dataset.

In [79]:
df_test=df_train_v2.merge(df_tra_agg_per_user, on='msno')
df_test

msno  is_churn  \
0       ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   
1       f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1   
2       zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1   
3       8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   
4       K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1   
...                                              ...       ...   
970955  OHnZbu+EVaP+vN7Z+OfT5OMcp90MWFZonmM0o3pb8FY=         0   
970956  S92bDK//uI6hk3u1vuApro0qJBQOToBozZ7lL1yTC+w=         0   
970957  eUa3xo16vpAjr43Cjlb6Kjf1NTILYyJIkBayJQdXWnw=         0   
970958  iZE41tbAQ65rJq60olkJT4BJzuUAYgQdfbEemXe/TTk=         0   
970959  oECkzJik4wKsbOEVY6UACLbmgM8qymFdb5cJaHrodY8=         0   

        total_num_transaction max_date_trans min_date_trans max_memb_exp  \
0                           6     2017-02-28     2015-03-31   2017-03-30   
1                           6     2017-03-11     2016-03-03   2017-04-11   
2                          20     2017-03-14     2015-08-19   2017-06-15   
3                          28     2016-02-25     2015-01-08   2018-01-08   
4                          32     2017-03-16     2015-01-24   2017-09-18   
...                       ...            ...            ...          ...   
970955                     27     2017-03-04     2015-01-03   2017-04-03   
970956                      9     2017-03-10     2016-07-10   2017-04-10   
970957                     20     2017-03-16     2015-08-16   2017-04-16   
970958                     21     2017-03-31     2015-02-14   2017-05-13   
970959                     27     2017-03-18     2015-01-17   2017-04-18   

        min_amount_paid  max_amount_paid  mean_amount_paid  \
0                   149              149        149.000000   
1                     0              180        134.500000   
2                     0              300        149.100000   
3                     0              149        143.678571   
4                    99              149        117.125000   
...                 ...              ...               ...   
970955                0              149        143.481481   
970956               99               99         99.000000   
970957               99               99         99.000000   
970958              149              149        149.000000   
970959              149              149        149.000000   

        median_amount_paid  ...  median_delta_trans_date  \
0                    149.0  ...                     31.0   
1                    149.0  ...                     34.0   
2                    149.0  ...                     32.0   
3                    149.0  ...                     17.0   
4                    114.0  ...                     30.0   
...                    ...  ...                      ...   
970955               149.0  ...                     31.0   
970956                99.0  ...                     31.0   
970957                99.0  ...                     31.0   
970958               149.0  ...                     31.0   
970959               149.0  ...                     31.0   

        mode_delta_trans_date  min_delta_amount_paid  max_delta_amount_paid  \
0                        31.0                    0.0                    0.0   
1                         2.0                 -149.0                  180.0   
2                        33.0                 -149.0                  300.0   
3                         0.0                 -149.0                    0.0   
4                        31.0                  -30.0                   30.0   
...                       ...                    ...                    ...   
970955                   31.0                    0.0                  149.0   
970956                   31.0                    0.0                    0.0   
970957                   31.0                    0.0                    0.0   
970958                   31.0                    0.0                    0.0   
970959 

In [74]:
df_user_all_info_agg_for_model=df_user_all_info_agg.merge(df_train_v2,on="msno")
df_user_all_info_agg_for_model

msno  city  bd  gender  \
0       +tJonkh+O1CA796Fm5X60UMOtB6POHAwPjbTRVl/EuU=     1   0     NaN   
1       yLkV2gbZ4GLFwqTOXLVHz0VGrMYcgBGgKZ3kj9RiYu8=     4  30    male   
2       I0yFvqMoNkM8ZNHb617e1RBzIS/YRKemHO7Wj13EtA0=    13  63    male   
3       OoDwiKZM+ZGr9P3fRivavgOtglTEaNfWJO4KaJcTTts=     1   0     NaN   
4       4De1jAxNRABoyRBDZ82U0yEmzYkqeOugRGVNIf92Xb8=     4  28  female   
...                                              ...   ...  ..     ...   
855131  dz8iQV2eZCBZo1DjKL7OifsiyzGf3xyMaCFDOJHVeyc=     1   0     NaN   
855132  OL6joplCKThE1pZD+AQDnj3kZiC45qzpndlDIRSyPQI=     1   0     NaN   
855133  O11rAa17ANM6DGMMVb9L2vwCQFv/S6SR3SDB1nGuOAc=     1   0     NaN   
855134  tyoRqQ72Y2oCcqvdaIxG8T0Hs9DZvuuM8IeTwhxtdBQ=     6  26  female   
855135  nWjH7glPkZ7jOVaCRwwjlpmp0T1hSWdv8hMJxiWCwKc=     1   0     NaN   

        registered_via  registration_init_time registration_init_time_as_dt  \
0                    7                20110914                   2011-09-14   
1                    9                20110916                   2011-09-16   
2                    9                20110918                   2011-09-18   
3                    7                20110918                   2011-09-18   
4                    9                20110920                   2011-09-20   
...                ...                     ...                          ...   
855131               7                20151020                   2015-10-20   
855132               7                20151020                   2015-10-20   
855133               7                20151020                   2015-10-20   
855134               7                20151020                   2015-10-20   
855135               7                20151020                   2015-10-20   

        activity_duration last_date_active  n_active_days  ...  \
0                     820       2017-03-31            763  ...   
1                     781       2017-03-31             18  ...   
2                     817       2017-03-30            354  ...   
3                     817       2017-03-28            295  ...   
4                     761       2017-03-31             37  ...   
...                   ...              ...            ...  ...   
855131                528       2017-03-31            108  ...   
855132                528       2017-03-31            396  ...   
855133                525       2017-03-31            273  ...   
855134                528       2017-03-31            417  ...   
855135                518       2017-03-26            188  ...   

        mode_delta_trans_date  min_delta_amount_paid  max_delta_amount_paid  \
0                        31.0                  -20.0                   20.0   
1                        31.0                    0.0                    0.0   
2                        31.0                    0.0                    0.0   
3                        31.0                    0.0                    0.0   
4                        31.0                    0.0                   31.0   
...                       ...                    ...                    ...   
855131                   31.0                    0.0                    0.0   
855132                   31.0                    0.0                    0.0   
855133                   31.0                    0.0                    0.0   
855134                   31.0                  -99.0                  149.0   
855135                   31.0                    0.0                    0.0   

        mode_delta_amount_paid  \
0                          0.0   
1                          0.0   
2                          0.0   
3                          0.0   
4                          0.0   
...                        ...   
855131                     0.0   
855132                     0.0   
855133                     0.0   
855134                     0.0   
855135                     0.0   

        duration_max_memb_exp_min_trans_date_effective  \
0

In [80]:
df_user_all_info_agg_for_model.to_csv("df_user_log__transfer__membership__aggregated_per_user__training_members.csv")